tests/sm_int/sm_int.c

```diff
@@ -1517,4 +1592,57 @@ TEST_FUNCTION(STATE_and_API)
     }
 }
 
+typedef struct WAIT_FOR_CANCEL_TAG
+{
+    SM_HANDLE sm;
+    volatile_atomic int32_t cancel_wait;
+} WAIT_FOR_CANCEL;
+
+static int waitForCancel(void* context)
+{
+    WAIT_FOR_CANCEL* cancel_wait_context = context;
+    ASSERT_IS_NOT_NULL(cancel_wait_context);
+    int32_t current_value = interlocked_add(&cancel_wait_context->cancel_wait, 0);
+    sm_exec_begin(cancel_wait_context->sm);
+    if (wait_on_address(&cancel_wait_context->cancel_wait, current_value, UINT32_MAX) != WAIT_ON_ADDRESS_OK)
+    {
+        ASSERT_FAIL("wait_on_address failed.");
+    }
+    sm_exec_end(cancel_wait_context->sm);
+    return 0;
+}
+
+static void cancellingCallback(void* context)
+{
+    volatile_atomic int32_t* address = context;
+    ASSERT_IS_NOT_NULL(address);
+    wake_by_address_single(address);

```

Comment:
>wake_by_address_single [](http://example.com/codeflow?start=4&length=22)

Should set the value first with `interlocked_exchange` or `InterlockedHL_SetAndWake` instead